pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -120,96 +119,157 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 				} else {
-					// This is the case of TCP or PASSTHROUGH.
-					tcpChainOpts := configgen.createGatewayTCPFilterChainOpts(builder.node, builder.push,
-						server, mergedGateway.GatewayNameForServer[server])
-					newFilterChainOpts = append(newFilterChainOpts, tcpChainOpts...)
-					for i := 0; i < len(tcpChainOpts); i++ {
-						newFilterChains = append(newFilterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolTCP})
+					// build http connection manager with TLS context, for HTTPS servers using simple/mutual TLS
+					// build listener with tcp proxy, with or without TLS context, for TCP servers
+					//   or TLS servers using simple/mutual/passthrough TLS
+					//   or HTTPS servers using passthrough TLS
+					// This process typically yields multiple filter chain matches (with SNI) [if TLS is used]
+					tcpFilterChainOpts := make([]*filterChainOpts, 0)
+					for _, server := range serversForPort.Servers {
+						if gateway.IsTLSServer(server) && gateway.IsHTTPServer(server) {
+							routeName := mergedGateway.TLSServerInfo[server].RouteName
+							// This is a HTTPS server, where we are doing TLS termination. Build a http connection manager with TLS context
+							tcpFilterChainOpts = append(tcpFilterChainOpts, configgen.createGatewayHTTPFilterChainOpts(builder.node, server.Port, server,
+								routeName, proxyConfig, istionetworking.TransportProtocolTCP))
+							newFilterChains = append(newFilterChains, istionetworking.FilterChain{
+								ListenerProtocol:   istionetworking.ListenerProtocolHTTP,
+								IstioMutualGateway: server.Tls.Mode == networking.ServerTLSSettings_ISTIO_MUTUAL,
+							})
+						} else {
+							// This is the case of TCP or PASSTHROUGH.
+							tcpChainOpts := configgen.createGatewayTCPFilterChainOpts(builder.node, builder.push,
+								server, mergedGateway.GatewayNameForServer[server])
+							tcpFilterChainOpts = append(tcpFilterChainOpts, tcpChainOpts...)
+							for i := 0; i < len(tcpChainOpts); i++ {
+								newFilterChains = append(newFilterChains, istionetworking.FilterChain{
+									ListenerProtocol: istionetworking.ListenerProtocolTCP,
+								})
+							}
+						}
 					}
+
+					opts.filterChainOpts = tcpFilterChainOpts
+				}
+			case istionetworking.TransportProtocolQUIC:
+				// Currently, we just assume that QUIC is HTTP/3 although that does not
+				// have to be the case (it is just the most common case now, in the future
+				// we will support more cases)
+				quicFilterChainOpts := make([]*filterChainOpts, 0)
+				for _, server := range serversForPort.Servers {
+					log.Debugf("buildGatewayListeners: creating QUIC filter chain for port %d(%s:%s)",
+						server.GetPort().GetNumber(), server.GetPort().GetName(), server.GetPort().GetProtocol())
+					routeName := mergedGateway.TLSServerInfo[server].RouteName
+					if mergedGateway.TLSToQUICServerRouteMap[routeName] != "" {
+						routeName = mergedGateway.TLSToQUICServerRouteMap[routeName]
```

Comment:
is `mergedGateway.TLSToQUICServerRouteMap[routeName]` ever `""` here?

Reply:
I originally wrote something and this remained. I changed the name and type to set. It is the set of routes on which H3 has to be advertised (by setting `alt-svc` header)